runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkDetachedRunnerResult.java

```diff
@@ -18,41 +18,119 @@
+
+  private JobClient jobClient;
+  private int jobCheckIntervalInSecs;
+
+  FlinkDetachedRunnerResult(JobClient jobClient, int jobCheckIntervalInSecs) {
+    this.jobClient = jobClient;
+    this.jobCheckIntervalInSecs = jobCheckIntervalInSecs;
+  }
 
   @Override
   public State getState() {
-    return State.UNKNOWN;
+    try {
+      return toBeamJobState(jobClient.getJobStatus().get());
+    } catch (InterruptedException | ExecutionException e) {
+      throw new RuntimeException("Fail to get flink job state", e);
+    }
   }
 
   @Override
   public MetricResults metrics() {
-    throw new UnsupportedOperationException("The FlinkRunner does not currently support metrics.");
+    return MetricsContainerStepMap.asAttemptedOnlyMetricResults(getMetricsContainerStepMap());
+  }
+
+  private MetricsContainerStepMap getMetricsContainerStepMap() {
+    try {
+      return (MetricsContainerStepMap)
+          jobClient
+              .getAccumulators()
+              .get()
+              .getOrDefault(FlinkMetricContainer.ACCUMULATOR_NAME, new MetricsContainerStepMap());
+    } catch (InterruptedException | ExecutionException e) {
+      LOG.warn("Fail to get flink job accumulators", e);
+      return new MetricsContainerStepMap();
+    }
   }
 
   @Override
   public State cancel() throws IOException {
-    throw new UnsupportedOperationException("Cancelling is not yet supported.");
+    try {
+      this.jobClient.cancel().get();
+    } catch (InterruptedException | ExecutionException e) {
+      throw new RuntimeException("Fail to cancel flink job", e);
+    }
+    return getState();
   }
 
   @Override
   public State waitUntilFinish() {
-    return State.UNKNOWN;
+    return waitUntilFinish(Duration.standardDays(TEN_YEAR_DAYS));
   }
 
   @Override
   public State waitUntilFinish(Duration duration) {
-    return State.UNKNOWN;
+    long start = System.currentTimeMillis();
+    long durationInMillis = duration.getMillis();
+    State state = State.UNKNOWN;
+    while ((System.currentTimeMillis() - start) < durationInMillis) {
+      state = getState();
+      if (state.isTerminal()) {
+        return state;
+      }
+      try {
+        Thread.sleep(jobCheckIntervalInSecs * 1000);
+      } catch (InterruptedException e) {
+        throw new RuntimeException(e);
+      }
+    }
+
+    if (state != null && !state.isTerminal()) {
+      LOG.warn("Job is not finished in {} seconds", duration.getStandardSeconds());
+    }
+    return state;
+  }
+
+  private State toBeamJobState(JobStatus flinkJobStatus) {
+    switch (flinkJobStatus) {
+      case CANCELLING:
+      case CREATED:
+      case INITIALIZING:
+      case FAILING:
+      case RECONCILING:
+      case RESTARTING:
+      case RUNNING:
+        return State.RUNNING;
+      case FINISHED:
+        return State.DONE;
+      case CANCELED:
+        return State.CANCELLED;
+      case FAILED:
+        return State.FAILED;
+      case SUSPENDED:
```

Comment:
should this return [STOPPED](https://beam.apache.org/releases/javadoc/2.1.0/org/apache/beam/sdk/PipelineResult.State.html#STOPPED) beam state instead, where the job is in paused state and not in a terminal state? Flink's [suspended](https://nightlies.apache.org/flink/flink-docs-stable/api/java/org/apache/flink/api/common/JobStatus.html#SUSPENDED) state also corresponds to paused state where the job may transition to [restarting state](https://nightlies.apache.org/flink/flink-docs-release-1.17/docs/internals/job_scheduling/)

Reply:
Thanks for the review, @kkdoon I have updated this.